clang-tools-extra/clang-tidy/modernize/UseDigitSeparatorCheck.cpp

```diff
@@ -0,0 +1,235 @@
+      OriginalLiteralString.substr(OriginalDotPosition + 1,
+                                   OriginalLiteralString.size());
+  int Precision = 0;
+  for (const char &Character : OriginalFractionalSubString) {
+    if (std::isdigit(Character)) {
+      Precision++;
+    }
+  }
+
+  // Get formatting literal text
+
+  // Get string representation of float value
+  std::ostringstream StringStream;
+  StringStream << std::fixed << std::setprecision(Precision)
+               << FloatValue.convertToDouble();
+  const std::string FloatString = StringStream.str();
+
+  // Get integer and fractional parts of float number
+  const std::string::size_type DotPosition = FloatString.find('.');
+  const std::string IntegerSubString = FloatString.substr(0, DotPosition);
+  std::string FractionalSubString =
+      FloatString.substr(DotPosition + 1, FloatString.size());
+  std::reverse(FractionalSubString.begin(), FractionalSubString.end());
+
+  // Get formatting literal text
+  const std::string FormatedIntegerSubString = getFormatedIntegerString(
+      IntegerSubString, llvm::APInt(128, std::stoll(IntegerSubString)));
+  std::string FormatedFractionalSubString = getFormatedIntegerString(
+      FractionalSubString, llvm::APInt(128, std::stoll(FractionalSubString)));
+  std::reverse(FormatedFractionalSubString.begin(),
+               FormatedFractionalSubString.end());
+
+  const std::string FormatedLiteralString =
+      FormatedIntegerSubString + '.' + FormatedFractionalSubString + Postfix;
+
+  return FormatedLiteralString;
+}
+
+std::string
+getFormatedScientificFloatString(const llvm::StringRef OriginalLiteralString) {
+  // Split string to mantissa and exponent
+  const char EChar =
+      OriginalLiteralString.str().find('E') != std::string::npos ? 'E' : 'e';
+  const std::string::size_type EPosition =
+      OriginalLiteralString.str().find(EChar);
+  const std::string SignSymbol =
+      OriginalLiteralString[EPosition + 1] == '-'
+          ? "-"
+          : (OriginalLiteralString[EPosition + 1] == '+' ? "+" : "");
+  const llvm::StringRef MantissaSubString =
+      OriginalLiteralString.substr(0, EPosition);
+  const llvm::StringRef ExponentSubString = OriginalLiteralString.substr(
+      EPosition + SignSymbol.size() + 1, OriginalLiteralString.size());
+
+  // Get formatting literal text
+  const std::string FormatedMantissaString = getFormatedFloatString(
+      MantissaSubString, llvm::APFloat(std::stod(MantissaSubString.str())));
+  const std::string FormatedExponentString = getFormatedIntegerString(
+      ExponentSubString, llvm::APInt(128, std::stoll(ExponentSubString.str())));
+  return FormatedMantissaString + EChar + SignSymbol + FormatedExponentString;
+}
+} // namespace
+
+namespace clang::tidy::modernize {
+
+void UseDigitSeparatorCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(integerLiteral().bind("integerLiteral"), this);
+  Finder->addMatcher(floatLiteral().bind("floatLiteral"), this);
```

Comment:
consider excluding system code. and consider adding some checks here, like size of int/float to exclude those were separator is not needed, or even exclude those where separator is already used.